### ISSUE TITLE
GPU budget for finetuning

### ISSUE THREAD

**Issue by BaohaoLiao** [allenai/open-instruct#12]
Hi,

may I ask what is the required GPU memory when using the same scripts as yours for the no_offload and offload ds config?

I use 4GPU with 48GB memory and a batch size of 1, but it fails for both offload and no_offload ds configs. BTW, I use the LoRA fine-tuning script.

**Comment by hamishivi** [comment]
Hi, I was able to finetune 7B fine on a single 80GB A100, so training over 4 gpus should work (esp with offload, since that offloads the model state). What size model are you trying to train?

**Comment by BaohaoLiao** [comment]
I use also use the 7B model.

**Comment by hamishivi** [comment]
This is using a max sequence length of 2048. Maybe it's an issue with PEFT and distributed training? We didn't test the lora script that much (since it wasn't used beyond some initial experiments).

Is it failing with a CUDA oom error, or something else?

**Comment by BaohaoLiao** [comment]
Yes, it’s a Cuda OOM error. I also try the full fine-tuning, similar error.

**Comment by hamishivi** [comment]
Hi, I ran the Lora finetuning successfully on 2 x 40GB A100s using the codebase (well, its made 10 gradient steps without OOMing). Did you make sure to change `export CUDA_VISIBLE_DEVICES=0,1` at the top of the script? My script looks like:

```bash
export CUDA_VISIBLE_DEVICES=0,1

MODEL_SIZE=7B
NUM_GPUS=2
BATCH_SIZE_PER_GPU=1
TOTAL_BATCH_SIZE=128
GRADIENT_ACC_STEPS=$(($TOTAL_BATCH_SIZE/$NUM_GPUS/$BATCH_SIZE_PER_GPU))
echo "Training llama model ${MODEL_SIZE} using $NUM_GPUS GPUs, $BATCH_SIZE_PER_GPU batch size per GPU, $GRADIENT_ACC_STEPS gradient accumulation steps"

# Lora training
accelerate launch \
    --mixed_precision bf16 \
    --num_machines 1 \
    --num_processes $NUM_GPUS \
    --use_deepspeed \
    --deepspeed_config_file ds_configs/stage3_no_offloading_accelerate.conf \
    open_instruct/finetune.py \
    --model_name_or_path ~/hf_llama/${MODEL_SIZE} \
    --use_flash_attn \
    --use_lora \
    --lora_rank 256 \
    --lora_alpha 256 \
    --lora_dropout 0.05 \
    --tokenizer_name ~/hf_llama/${MODEL_SIZE} \
    --use_slow_tokenizer \
    --train_file ~/oasst1_data.jsonl \
    --max_seq_length 2048 \
    --preprocessing_num_workers 16 \
    --per_device_train_batch_size $BATCH_SIZE_PER_GPU \
    --gradient_accumulation_steps $GRADIENT_ACC_STEPS \
    --learning_rate 2e-5 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0. \
    --num_train_epochs 2 \
    --output_dir output/alpaca_${MODEL_SIZE}_lora/ \
    --save_merged_lora_model \
    --with_tracking \
    --report_to tensorboard \
    --logging_steps 1 &&

python open_instruct/merge_lora.py \
    --base_model_name_or_path /net/nfs.cirrascale/allennlp/yizhongw/hf_llama_models/${MODEL_SIZE} \
    --lora_model_name_or_path output/alpaca_${MODEL_SIZE}_lora/
    ```

**Comment by BaohaoLiao** [comment]
I might know the reason. Does `--use_flash_attn ` really matter? My GPU is NVIDIA RTX A6000. It doesn't support flash attention. Could you try without it?

**Comment by hamishivi** [comment]
Ah, you're totally right - I get OOM without flash attention. Flash attention does matter a lot, especially if you're trying to train with long sequences, so it makes sense that reducing the sequence length matters. Training with 1024 sequence length isn't too bad, and it doesn't hurt performance much I think (although I don't have exact numbers to back this up).

**Comment by BaohaoLiao** [comment]
Thank you very much for your patient help!
